File: grimdb/sdk/csharp/src/Models/AuditEvent.cs
using System.Text.Json.Serialization;

namespace Grimlocker.Models;

public sealed record AuditEvent
{
    [JsonPropertyName("timestamp")]
    public long Timestamp { get; init; }

    [JsonPropertyName("level")]
    public string Level { get; init; } = "";

    [JsonPropertyName("module")]
    public string Module { get; init; } = "";

    [JsonPropertyName("message")]
    public string Message { get; init; } = "";

    [JsonPropertyName("subject_id")]
    public string? SubjectId { get; init; }
}

File: grimdb/sdk/php/src/Models/AuditEvent.php
<?php declare(strict_types=1);
namespace Grimlocker\Models;

class AuditEvent
{
    public function __construct(
        public readonly int     $timestamp = 0,
        public readonly string  $level     = '',
        public readonly string  $module    = '',
        public readonly string  $message   = '',
        public readonly ?string $subjectId = null,
    ) {}

    public static function fromArray(array $data): self
    {
        return new self(
            $data['timestamp']  ?? 0,
            $data['level']      ?? '',
            $data['module']     ?? '',
            $data['message']    ?? '',
            $data['subject_id'] ?? null,
        );
    }
}
